lib/std/io/auto_indenting_stream.zig

```diff
@@ -0,0 +1,135 @@
+const std = @import("../std.zig");
+const io = std.io;
+const mem = std.mem;
+const assert = std.debug.assert;
+
+pub fn AutoIndentingStream(comptime indent_delta: u8, comptime OutStreamType: type) type {
+    return struct {
+        const Self = @This();
+        pub const Error = OutStreamType.Error;
+        pub const OutStream = io.Writer(*Self, Error, write);
+
+        out_stream: *OutStreamType,
+        current_line_empty: bool = true,
+        indent_stack: [255]u8 = undefined,
+        indent_stack_top: u8 = 0,
+        indent_one_shot_count: u8 = 0, // automatically popped when applied
+        applied_indent: u8 = 0, // the most recently applied indent
+        indent_next_line: u8 = 0, // not used until the next line
+
+        pub fn init(out_stream: *OutStreamType) Self {
+            return Self{ .out_stream = out_stream };
+        }
+
+        pub fn writer(self: *Self) OutStream {
+            return .{ .context = self };
+        }
+
+        pub fn write(self: *Self, bytes: []const u8) Error!usize {
+            if (bytes.len == 0)
+                return @as(usize, 0);
+
+            try self.applyIndent();
+            return self.writeNoIndent(bytes);
+        }
+
+        fn writeNoIndent(self: *Self, bytes: []const u8) Error!usize {
+            try self.out_stream.outStream().writeAll(bytes);
+            if (bytes[bytes.len - 1] == '\n')
+                self.resetLine();
+            return bytes.len;
+        }
+
+        pub fn insertNewline(self: *Self) Error!void {
+            _ = try self.writeNoIndent("\n");
+        }
+
+        fn resetLine(self: *Self) void {
+            self.current_line_empty = true;
+            self.indent_next_line = 0;
+        }
+
+        /// Insert a newline unless the current line is blank
+        pub fn maybeInsertNewline(self: *Self) Error!void {
+            if (!self.current_line_empty)
+                try self.insertNewline();
+        }
+
+        /// Push default indentation
+        pub fn pushIndent(self: *Self) void {
+            // Doesn't actually write any indentation. Just primes the stream to be able to write the correct indentation if it needs to.
+            self.pushIndentN(indent_delta);
+        }
+
+        /// Push an indent of arbitrary width
+        pub fn pushIndentN(self: *Self, n: u8) void {
+            assert(self.indent_stack_top < std.math.maxInt(u8));
```

Comment:
so `zig fmt` will panic/exhibit undefined behaviour if you have > 255 levels of indentation?

Reply:
Yeah, I guess that could possibly maybe happen with generated code. Worth bumping to a u32?

Reply:
Though looking again, if we want to store a large number of indentations we would then need to dynamically allocate and pass in an allocator.

Reply:
Can be used LinearFifo here? With `{ .Static }` option.